### PR TITLE
Add delete item endpoint

### PR DESCRIPTION
**Pull request by samstradling**
Adds the endpoint to delete an item on a list:

```http
DELETE /lists/<list_id>/items/<item_id>
```

### CODE REVIEW

File: db/db.go [hunk 1]
```diff
@@ -6,6 +6,7 @@ import (
 
 // DB - interface for talking to the database
 type DB interface {
+	DeleteItem(*string, *string) error
```
Comment (alexNaylorTeece) [inline db/db.go:9]: why pointers?

File: db/deleteItem.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+)
+
+func (d *dynamoDB) DeleteItem(listID *string, itemID *string) error {
+	input := &dynamodb.DeleteItemInput{
+		Key: map[string]*dynamodb.AttributeValue{
+			"ListId": {S: listID},
+			"Id":     {S: itemID},
+		},
+		TableName: aws.String(d.conf.TableNames.Items),
```
Comment (alexNaylorTeece) [inline db/deleteItem.go:15]: Will this be fine if `Items` is an empty string?

Comment (alexNaylorTeece) [reply db/deleteItem.go:15]: Assume that dynamo will just fail? Would it be better to catch that early (before we make the request)?
Up to you

File: db/deleteItem.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+)
+
+func (d *dynamoDB) DeleteItem(listID *string, itemID *string) error {
+	input := &dynamodb.DeleteItemInput{
+		Key: map[string]*dynamodb.AttributeValue{
+			"ListId": {S: listID},
```
Comment (alexNaylorTeece) [inline db/deleteItem.go:12]: If listID and itemID weren't pointers you could use `&listID` and `&itemID` here
That way you don't have to worry about them maybe being nil

File: db/deleteItem.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+			"Id":     {S: itemID},
+		},
+		TableName: aws.String(d.conf.TableNames.Items),
+	}
+
+	_, err := d.session.DeleteItem(input)
+
+	if err != nil {
```
Comment (alexNaylorTeece) [inline db/deleteItem.go:20]: Could use the error switching that I put in the update PR here https://github.com/mount-joy/thelist-lambda/pull/9/files#r516787910

File: db/deleteItem_test.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDeleteItem(t *testing.T) {
+	listID := "474c2Fff7"
+	itemID := "b6cf642d"
+	tableName := "items-table"
+	tests := []struct {
+		name        string
+		outputErr   error
```
Comment (alexNaylorTeece) [inline db/deleteItem_test.go:18]: Not clear this is for mock

File: db/deleteItem.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == dynamodb.ErrCodeConditionalCheckFailedException {
+				return NewError(ErrorNotFound)
```
Comment (alexNaylorTeece) [inline db/deleteItem.go:23]: For a delete, surely a not found should just return success?

File: handlers/deleteitem/deleteItem.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+// New returns an instance of deleteItem satisfying the RouteHandler interface
+func New() iface.RouteHandler {
+	return &deleteItem{
+		db: db.DynamoDB(),
+	}
+}
+
+// Match returns true if this RouteHandler should handle this request
+func (g *deleteItem) Match(request events.APIGatewayV2HTTPRequest) bool {
+	var re = regexp.MustCompile(`^/lists/([\w-]+)/items/([\w-]+)/?$`)
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem.go:28]: Comment to say what it should match would be helpful

File: handlers/deleteitem/deleteItem.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+// New returns an instance of deleteItem satisfying the RouteHandler interface
+func New() iface.RouteHandler {
+	return &deleteItem{
+		db: db.DynamoDB(),
+	}
+}
+
+// Match returns true if this RouteHandler should handle this request
+func (g *deleteItem) Match(request events.APIGatewayV2HTTPRequest) bool {
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem.go:27]: Think this being called g is a hangover from a copy and paste?

File: handlers/deleteitem/deleteItem.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	var re = regexp.MustCompile(`^/lists/([\w-]+)/items/([\w-]+)/?$`)
+	return request.RequestContext.HTTP.Method == "DELETE" && re.MatchString(request.RequestContext.HTTP.Path)
+}
+
+// Handle handles this request and returns the response and status code
+func (g *deleteItem) Handle(request events.APIGatewayV2HTTPRequest) (interface{}, int) {
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem.go:33]: ditto g

File: handlers/deleteitem/deleteItem.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+			}
+		}
+		log.Printf("Error: %s", err.Error())
+		return nil, http.StatusInternalServerError
+	}
+
+	return nil, http.StatusOK
+}
+
+func getIDs(path string) (*string, *string, error) {
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem.go:54]: Wonder if a function with the signature: `func getElementFromPath(elementNumber int, path string) (string, error)`
Could be used for all routes, but probably appropriate for a different PR

File: handlers/deleteitem/deleteItem.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+	parts := strings.SplitN(path, "/", 6)
+	if len(parts) < 5 {
+		return nil, nil, fmt.Errorf("Unable to match path: %s", path)
+	}
+	return &parts[2], &parts[4], nil
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem.go:59]: Still confused why always using pointers to the string

File: handlers/deleteitem/deleteItem.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+// Handle handles this request and returns the response and status code
+func (g *deleteItem) Handle(request events.APIGatewayV2HTTPRequest) (interface{}, int) {
+	listID, itemID, err := getIDs(request.RequestContext.HTTP.Path)
+	if err != nil {
+		log.Printf("Error: %s", err.Error())
+		return nil, http.StatusBadRequest
+	}
+
+	err = g.db.DeleteItem(listID, itemID)
+	if err != nil {
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem.go:41]: As before with using a switch to reduce indenting

File: handlers/deleteitem/deleteItem.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+		if aerr, ok := err.(db.Error); ok {
+			if aerr.ErrorType == db.ErrorNotFound {
+				return nil, http.StatusNotFound
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem.go:44]: Status not found on a delete feels odd

File: handlers/deleteitem/deleteItem_test.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dbMocked := &testhelpers.MockDB{}
+			dbMocked.Test(t)
+			defer dbMocked.AssertExpectations(t)
+
+			input := testhelpers.CreateAPIGatewayV2HTTPRequest(tt.path, tt.method, "")
+			d := deleteItem{db: dbMocked}
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem_test.go:102]: Don't think you actually need to set the `db` for the match function?

File: handlers/deleteitem/deleteItem_test.go [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+	}
+}
+
+func TestDeleteItemHandle(t *testing.T) {
+	tests := []struct {
+		name               string
+		path               string
+		listID             string
+		itemID             string
+		output             *data.Item
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem_test.go:116]: Not clear this is for the mock

File: handlers/deleteitem/deleteItem_test.go [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+		outputErr          error
+		expectedStatusCode int
+		shouldCallDB       bool
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem_test.go:119]: As in the update PR about maybe using a struct to eliminate the need for this
https://github.com/mount-joy/thelist-lambda/pull/9/files#r516857435

File: handlers/deleteitem/deleteItem_test.go [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+			expectedStatusCode: 500,
+			shouldCallDB:       true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dbMocked := &testhelpers.MockDB{}
+			dbMocked.Test(t)
+			if tt.shouldCallDB {
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem_test.go:175]: As before in the update PR about where this should be
https://github.com/mount-joy/thelist-lambda/pull/9/files#r516857435

File: handlers/deleteitem/deleteItem_test.go [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+				defer dbMocked.AssertExpectations(t)
+			}
+
+			dbMocked.
+				On("DeleteItem", &tt.listID, &tt.itemID).
+				Return(tt.output, tt.outputErr)
```
Comment (alexNaylorTeece) [inline handlers/deleteitem/deleteItem_test.go:181]: As before https://github.com/mount-joy/thelist-lambda/pull/9/files#r516855258

**Review by alexNaylorTeece — CHANGES_REQUESTED** [review]

File: handlers/deleteitem/deleteItem.go (hunk 10 above)
Comment (samstradling) [reply handlers/deleteitem/deleteItem.go:54]: Yes I agree, let's fix it outside of this PR though

File: db/deleteItem.go (hunk 4 above)
Comment (samstradling) [reply db/deleteItem.go:20]: I couldn't get this to work in the other PR but if I'm able to fix it there I'll update it here too

Comment (alexNaylorTeece) [reply db/deleteItem.go:25]: https://play.golang.org/p/zqM53h68Cdz should help

File: handlers/deleteitem/deleteItem.go (hunk 12 above)
Comment (samstradling) [reply handlers/deleteitem/deleteItem.go:42]: We ended up removing this check because of your other comment so no need 😄

File: db/deleteItem.go [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+	if len(tableName) == 0 {
+		panic("Items table name not set")
+	}
+
+	input := &dynamodb.DeleteItemInput{
+		Key: map[string]*dynamodb.AttributeValue{
+			"ListId": {S: &listID},
+			"Id":     {S: &itemID},
+		},
+		TableName: aws.String(d.conf.TableNames.Items),
```
Comment (alexNaylorTeece) [inline db/deleteItem.go:20]: You've looked this up and set it to `tableName` on line 10. Might as well use `tableName` here

File: db/deleteItem.go [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+		break
+	case awserr.Error:
+		if e.Code() == dynamodb.ErrCodeConditionalCheckFailedException {
+			return nil
+		}
+	default:
+		return err
+	}
+
+	return err
```
Comment (alexNaylorTeece) [inline db/deleteItem.go:36]: You know err is nil if you came here. Might as well be explicit and have `return nil`

**Review by alexNaylorTeece — APPROVED** [review]
Nice. Couple of optionals

File: db/deleteItem.go (hunk 19 above)
Comment (samstradling) [reply db/deleteItem.go:20]: yep sorry, this was a copy and paste error

**Pull request merged**